week 6 & 7& 8

bishoy-magdy opened the pull request:
# What has been done

- Added the feature of fetching the message data from imap.
- Installed the email parser.
- Added the feature of parsing the HTML message and extracting the body using DOMParser.
- Added the feature of marking the converted message in the plugin without editing the message or adding flags.
- Added the feature to save the last change on the screen (main screen) when the plugin is closed.
- Created email parser class and added the unit test for the email parser class.
- Added the ability to convert email notes to text.
- Created the class responsible for posting notes to Joplin.
- Added inline attachments to note.
-  Added the feature of converting emails to notes without logging in.


# Demo 

A demo example of the latest feature that was added

https://user-images.githubusercontent.com/58605547/183949858-a0b5fe9e-9e19-40a7-90b0-e91ed6dd1a31.mp4

bishoy-magdy committed:
week6,7,8

roman-r-m commented on src/core/emailParser.ts:
```diff
@@ -0,0 +1,36 @@
+
+            return (this.emailContent);
+        } catch (err) {
+            throw err;
```
I think doing this is equivalent to not using try/catch at all.

roman-r-m replied:
To be clear - I'm not saying you should or should not catch errors here, but certainly this is just pointless, no?

bishoy-magdy replied:
You are right, it's pointless. I forgot to remove it. I only used it to test and know the different types of errors from the email parser in the external JS file.

roman-r-m commented on src/core/imap.ts:
```diff
@@ -51,7 +59,7 @@ export class IMAP {
     }
 
-    search(criteria) {
+    search(criteria): Promise<number[]> {
```
Not a huge deal but you're code style is a bit inconsistent, here you annotate the return type, for some other functions in the same class you don't.

bishoy-magdy replied:
Okay, I have now refacted the imap file and installed @types/imap to provide the imap function interfaces.

 I noticed that I did not annotate the functions in the standard way. Now I annotated functions that have a promise function inside the promise

roman-r-m commented on src/core/imap.ts:
```diff
@@ -83,6 +91,81 @@ export class IMAP {
+            }
+
+            try {
+                const value = await joplin.settings.value(CONVERTED_MESSAGES);
```
This setting will grow over time, could become a performance bottleneck at some point.

bishoy-magdy replied:
I came up with a better solution: get the converted message and store it in a variable, then query that variable to see if there is a new message, and when there is, the values are updated. This means that the converted message in Joplin is not only updated until there is a new message(If you have another solution please tell me).

roman-r-m replied:
Not sure what you mean by variable here? You need to persists this information between restarts somehow.

It seems to me that the easiest way to store this information is to use the email account itself - you either delete messages you've converted, mark them read, or move to another folder.
Or you could only store the date of the last seen email and only process newer. This way you store it on the client but at least it's just one number per mailbox.

bishoy-magdy replied:
indeed, the thought came to my mind to store this information by using the email account itself, but there is no standard way to mark the message. To be more clear, some email providers allow you to mark messages while others don't.

But I found a solution One of the flags that are standard for all email providers is the '**SEEN**' flag. Every message is flagged as '**UNSEEN**' if the message is new or the user marks it as '**UNSEEN**'.

So I will add an additional condition in the search, which is to search for the messages that are "from" a specific email and marked as "**UNSEEN**" and then mark the messages as "**SEEN**" after fetching them.

I think this solution is better than deleting. and the user can mark as an "**unread**" message any number of messages in their email provider.

Sorry for the delay in suggesting this solution. It took me a lot of time to test this solution across different email providers.

bishoy-magdy replied:
example:

https://user-images.githubusercontent.com/58605547/185809305-565da1bb-046c-46cf-975e-45d0fcc75015.mp4

roman-r-m replied:
Just to confirm - you'll add this in another pr, right?

roman-r-m commented on src/core/postAttachments.ts:
```diff
@@ -0,0 +1,52 @@
+import {sep} from 'path';
+import joplin from 'api';
+import {Attachment} from '../model/attachment.model';
+import {AttachmentProperties} from '../model/attachmentProperties.mode';
```
a typo? `mode` vs `model` on the previous line

bishoy-magdy replied:
> a typo?

Yes, I fixed it.

bishoy-magdy replied:
> a typo?
Yes, I fixed it.

roman-r-m commented on src/core/postAttachments.ts:
```diff
@@ -0,0 +1,52 @@
+const fs = joplin.require('fs-extra');
+
+export class Attachments {
+    tempFolder: string = `${tmpdir}${sep}joplin-email-plugin${sep}`;
```
consider using `path.join` instead

roman-r-m replied:
Just noticed, you have a class `TempFolder` but still create the path manually here. Why not add a method to `TempFolder` instead? This way the path will only be defined in one place

bishoy-magdy replied:
> consider using path.join instead

Done

> Just noticed, you have a class TempFolder but still create the path manually here. Why not add a method to TempFolder instead? This way the path will only be defined in one place

It's not for getting the temp path, it is for deleting previous attachment folders and creating new temp folders.
I renamed the class to 'setupTempFolder' to be more clear, and this class is called in the init of the plugin.

roman-r-m replied:
The idea is you want to have the logic of getting the temp path only existing in one place. At the moment you have it duplicated here and in `setupTempFolder`.
Obviously in this case it's not a huge deal as this piece of code is very simple and is unlikely to change.

roman-r-m commented on src/core/postAttachments.ts:
```diff
@@ -0,0 +1,52 @@
+                    ],
+                );
+
+                attachmentsProp.push({contentId: contentId, id: resource.id});
```
might be worth cleaning up the temp file afterwards

bishoy-magdy replied:
I tried deleting folders in this place, but it will slow down when there are a lot of messages that you want to convert because it will create and delete and so on. What I did is every time the plugin starts, it deletes the old attachments and starts creating a new temporary folder.

roman-r-m replied:
Could potentially be a security risk, say, if you run the plugin in Joplin portable from a USB drive, your emails will be left on the computer.

PackElend replied:
can't the delete job be a self-running/isolated background job? 
The "old" temp folder could be renamed, a new one created, old, now renamed, deleted in the background

bishoy-magdy replied:
> Could potentially be a security risk, say, if you run the plugin in Joplin portable from a USB drive, your emails will be left on the computer.

Okay, I refacted the function to this mechanism.

 - Create a temp folder
- - Post all emails including attachments to Joplin.
- Remove a temp folder

bishoy-magdy replied:
> can't the delete job be a self-running/isolated background job?
> The "old" temp folder could be renamed, a new one created, old, now renamed, deleted in the background

It's a sync function, which means I can't go to the next line after finishing this function(delete a temp folder).

roman-r-m replied:
You could in theory use a web worker (or whatever the equivalent of threads in the JS world is called) but this looks like an overkill.

bishoy-magdy replied:
> You could in theory use a web worker (or whatever the equivalent of threads in the JS world is called) but this looks like an overkill.

This looks very good since we can add all the notes and attachments posting functions in the other JS thread.
I will read more about web worker. and I will definitely add it if I notice its presence will be important.

roman-r-m commented on src/core/postNote.ts:
```diff
@@ -0,0 +1,146 @@
+        const joplinFolders = await joplin.data.get(['folders']);
+
+        // Search by folder title
+        const joplinFoldersTitles = joplinFolders.items.map((joplinFolder)=> joplinFolder.title);
```
It is possible to have multiple folders with the same name because they can be nested, so this logic may not always work correctly.
However, I don't see an easy fix right now. Definitely something to keep in mind though.

bishoy-magdy replied:
I fixed it by searching by id and showing the path of folders in options.

https://user-images.githubusercontent.com/58605547/185251175-d45affa9-369d-4982-ac89-14d36e6f3cff.mp4

roman-r-m commented on src/core/postNote.ts:
```diff
@@ -0,0 +1,146 @@
+        this.emailHtmlBody = new DOMParser().parseFromString(emailContent.html, 'text/html').body.innerHTML;
+
+        // Converting email to (html, markdown, text) depends on what export type is in the plugin settings.
+        await this.converter();
```
Maybe it should be called `convert` instead?

bishoy-magdy replied:
Done :)

roman-r-m commented on src/core/postNote.ts:
```diff
@@ -0,0 +1,146 @@
+        });
+
+        for (let i = 0; i < emailFolders.length; i++) {
+            const folder = emailFolders[i];
```
Do I read this correctly that it creates a copy of the email in multiple folders? Tbh, I struggle to imagine why this may be needed...

bishoy-magdy replied:
I deleted it because we don't need it now, but this function takes an array of the folders that you want the note in and for each folder, creates the note and assigns it the tags.

roman-r-m commented on src/core/postNote.ts:
```diff
@@ -0,0 +1,146 @@
+
+        for (let i = 0; i < emailFolders.length; i++) {
+            const folder = emailFolders[i];
+            this.noteBody['parent_id'] = folder.id;
```
Looking at this line it's probably more than just a note body, more like a complete note object?

bishoy-magdy replied:
I'm confused about what I should name this object. I changed it to noteProp if you have another name welcome.

roman-r-m replied:
Just note? Up to you though

bishoy-magdy committed:
update -- week6,7,8

roman-r-m commented on src/core/postNote.ts:
```diff
@@ -50,47 +36,36 @@ export class PostNote {
-        });
-
         const emailTags = joplinTags.items.filter((e)=>{
             return this.tags.includes(e.title);
```
I can't find how these tags are set right now, but I think this will stop working if a tag is renamed. Best to use IDs as you did with folder.

I guess it's not a concern for manual one-time import but for monitoring mailbox it is.

bishoy-magdy replied:
I moved this code to the 'addTags' function and fixed it so that if 'has_more' in Joplin tags is true, this function now takes a set of tags and creates all tags that are not inside Joplin tags, and it returns all mentioned tags with all properties. 

If something is not clear, please tell me.

bishoy-magdy committed:
update -- week6,7,8

removed the feature of local marking messages, added the feature of marking messages using the email providers, fixed If 'has_more' is true in Joplin tags or folders, refacted TempFolder mechanism

roman-r-m commented on src/core/imap.ts:
```diff
@@ -165,28 +143,30 @@ export class IMAP {
                         this.pending = true;
-                        const data = await this.fetchAll(newMessages, {bodies: ''});
 
+                        setupTempFolder = new SetupTempFolder();
```
A minor thing but there is no reason to create a new instance on every iteration.
In fact, you don't need to create any instances of this class at all - all method can be made static

bishoy-magdy replied:
OK, it will be changed in the next PR.

roman-r-m commented on src/core/imap.ts:
```diff
@@ -165,28 +143,30 @@ export class IMAP {
-                        await this.updateConvertedMessages(newMessages);
-                        this.uids.push(...newMessages);
-
+                        setupTempFolder.removeTempFolder();
```
Not sure if it will be any faster but you don't need to delete the whole folder, only specific files corresponding to already processed emails.

bishoy-magdy replied:
Deleting an entire file is a little faster than looping through each file and deleting it.

>  only specific files corresponding to already processed emails.


This is actually only what will be in the file, which means that the whole file will be deleted, and I also want to leave no trace after doing what is required.

I tested it by converting more than 100 eml files. Some of them contain large files containing videos and images, and the conversion process works fine.

roman-r-m commented:
Overall looks good, I'm going to merge despite a couple of small comments. Feel free to ignore those or address in subsequent PRs.

bishoy-magdy commented:
> Overall looks good, I'm going to merge despite a couple of small comments. Feel free to ignore those or address in subsequent PRs.

Big thanks for this code review. In parallel, I made some changes for the purpose of improvement for some code that will be in the next PR.
